twister: improve logging of cmake commands

mbolivar-nordic opened the pull request:
Blocked by https://github.com/zephyrproject-rtos/zephyr/pull/49288

With verbose output enabled, twister logs the cmake arguments as a
python array whenever it invokes cmake to generate a build system.

This is clunky output for manual reproduction, because it forces the
developer to manually join the arguments together for consumption by
the shell.

As an enhancement for this use case, use the standard library
shlex.join() function to produce a string that can be copy/pasted
directly into the shell. We should use this function instead of
str.join, because it correctly handles things like shell quoting for
options with spaces in them.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
twister: improve logging of cmake commands

With verbose output enabled, twister logs the cmake arguments as a
python array whenever it invokes cmake to generate a build system.

This is clunky output for manual reproduction, because it forces the
developer to manually join the arguments together for consumption by
the shell.

As an enhancement for this use case, use the standard library
shlex.join() function to produce a string that can be copy/pasted
directly into the shell on POSIX platforms.

We should use this function instead of str.join, because it correctly
handles things like shell quoting for options with spaces in them.

This function is not available on Windows, so we fall back on behavior
that is similar to the old one there.

The main difference on Windows is that the output now includes the
path to cmake as well, and not just its arguments. (This is the same
on POSIX, and is intended to help debug if multiple cmakes are
installed on the same machine.)

We are running cmake from a couple of different places, so doing this
cleanly requires adding a shared module with the platform abstraction.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

carlescufi reviewed (dismissed):
Thanks so much for this!

stephanosio commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -262,12 +263,12 @@ def run_cmake(self, args=""):
         cmd = [cmake] + cmake_args
         kwargs = dict()
 
+        logger.debug("Calling cmake: %s", shlex.join(cmd))
```
Looks like `shlex.join` is a new function in Python 3.8.

It does not really matter anyway since we are going to be updating the minimum required Python version to 3.8 (though, strictly speaking this PR should be merged after that is done).

mbolivar-nordic replied:
>  this PR should be merged after that is done

Thank you for catching that; updated accordingly.

hakehuang commented:
shlex is not support in windows platform, so shall we protect the usage in os? besides, we need add shlex to requirements.txt

gopiotr commented:
@hakehuang @mbolivar-nordic 

> shlex is not support in windows platform

It depends what you mean by "not support".

There is no problem with "use" `shlex` on Windows OS. By this I understand, that you can import this package and then call `shlex.join()` function and your program do not rise any error (I verified it).

But there is a problem with what `shlex.join()` produce. The command created by it cannot be literally copied and pasted into Command Prompt or Powershell. For example let's consider following Twister call:

```
python scripts\twister --ninja -vv -p nrf52840dk_nrf52840 -T samples\hello_world
```

The list of arguments passed to `subprocess.Popen` is as following:
```
['C:\\Program Files\\CMake\\bin\\cmake.EXE', '-BC:\\zephyrproject\\zephyr\\twister-out\\nrf52840dk_nrf52840\\samples\\hello_world\\sample.basic.helloworld', '-SC:\\zephyrproject\\zephyr\\samples\\hello_world', '-DTC_RUNID=035fcf741557e5a833ab4082c5df73a2', '-DEXTRA_CFLAGS=-Werror', '-DEXTRA_AFLAGS=-Werror -Wa,--fatal-warnings', '-DEXTRA_LDFLAGS=-Wl,--fatal-warnings', '-DEXTRA_GEN_DEFINES_ARGS=--edtlib-Werror', '-GNinja', '-DBOARD=nrf52840dk_nrf52840']
```
And command created by `shlex.join()` is as follow:
```
'C:\Program Files\CMake\bin\cmake.EXE' '-BC:\zephyrproject\twister-out\nrf52840dk_nrf52840\samples\hello_world\sample.basic.helloworld' '-SC:\zephyrproject\zephyr\samples\hello_world' -DTC_RUNID=3db30869200a65656b06938257f83093 -DEXTRA_CFLAGS=-Werror '-DEXTRA_AFLAGS=-Werror -Wa,--fatal-warnings' -DEXTRA_LDFLAGS=-Wl,--fatal-warnings -DEXTRA_GEN_DEFINES_ARGS=--edtlib-Werror -GNinja -DBOARD=nrf52840dk_nrf52840
```
And such command cannot be run in `Command Prompt`, because it do not accept single quotes `'` - they have to be changed to double ones `"` like there:
```
"C:\Program Files\CMake\bin\cmake.EXE" "-BC:\zephyrproject\twister-out\nrf52840dk_nrf52840\samples\hello_world\sample.basic.helloworld" "-SC:\zephyrproject\zephyr\samples\hello_world" -DTC_RUNID=3db30869200a65656b06938257f83093 -DEXTRA_CFLAGS=-Werror "-DEXTRA_AFLAGS=-Werror -Wa,--fatal-warnings" -DEXTRA_LDFLAGS=-Wl,--fatal-warnings -DEXTRA_GEN_DEFINES_ARGS=--edtlib-Werror -GNinja -DBOARD=nrf52840dk_nrf52840
```
Other problem is when command created by `shlex.join()` will be executed by `Powershell`. There command has to start with `&` sign to call cmake exe file properly. Additionally `-DEXTRA_LDFLAGS=-Wl,--fatal-warnings` argument has to be put into quotes because it consists comma, what is a problem for Powershell. So, in the end the proper command for `Powershell` looks like this:
```
& 'C:\Program Files\CMake\bin\cmake.EXE' '-BC:\zephyrproject\twister-out\nrf52840dk_nrf52840\samples\hello_world\sample.basic.helloworld' '-SC:\zephyrproject\zephyr\samples\hello_world' -DTC_RUNID=3db30869200a65656b06938257f83093 -DEXTRA_CFLAGS=-Werror '-DEXTRA_AFLAGS=-Werror -Wa,--fatal-warnings' '-DEXTRA_LDFLAGS=-Wl,--fatal-warnings' -DEXTRA_GEN_DEFINES_ARGS=--edtlib-Werror -GNinja -DBOARD=nrf52840dk_nrf52840
```

Taking it all into consideration I wondering what is better option for Windows users: stay with list of arguments (`[...]`) or give them command produced by `shlex.join()` which will not work without some modifications... @mbolivar-nordic what do you think about it?

gopiotr commented:
BTW I also wondering if this is really problem for Windows users because when I run Twister via `Command Prompt` or `Powershell`, then Twister doesn't print logs from logger from spawned subprocesses where all building logs are printed (which contains CMake logs too). So, as far as this problem will not be solved, then Windows users will not have a chance to even see CMake command created by `shlex.join()` or in any other way.

mbolivar-nordic commented:
Ignore the latest push; it was inadvertent. I'll modify this to keep the change on POSIX only.

mbolivar-nordic commented:
> So, as far as this problem will not be solved, then Windows users will not have a chance to even see CMake command created by `shlex.join()` or in any other way.

@gopiotr I consider this a separate problem not in the scope of this PR. I modified the logging so that if this is ever fixed, windows users will see the list of command and its arguments when cmake is run, while POSIX users keep the copy/pastable string.

@hakehuang there is no need to add shlex to requirements.txt. It is a standard library module.

I also updated the PR to catch the situation where we run cmake scripts, not running cmake to generate a build system. I can't find anywhere else that we run cmake in twister.

gopiotr reviewed (approved):
@mbolivar-nordic - thank you a lot for your work!
